controller/controller_server/csi_controller_server.py

```diff
@@ -318,6 +303,79 @@ def ListVolumes(self, request, context):
+            return csi_pb2.CreateSnapshotResponse()
+
+        source_volume_id = request.source_volume_id
+        logger.info("Snapshot base name : {}. Source volume id : {}".format(request.name, source_volume_id))
+        secrets = request.secrets
+        user, password, array_addresses = utils.get_array_connection_info_from_secret(secrets)
+        try:
+            _, vol_id = utils.get_volume_id_info(source_volume_id)
+            with ArrayConnectionManager(user, password, array_addresses) as array_mediator:
+                logger.debug(array_mediator)
+                # TODO: CSI-1358 - remove try/except
+                try:
+                    snapshot_name = self._get_snapshot_name(request, array_mediator)
+                except controller_errors.IllegalObjectName as ex:
+                    context.set_details(ex.message)
+                    context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
+                    return csi_pb2.CreateSnapshotResponse()
+
+                volume_name = array_mediator.get_volume_name(vol_id)
+                logger.info("Snapshot name : {}. Volume name : {}".format(snapshot_name, volume_name))
+                snapshot = array_mediator.get_snapshot(snapshot_name)
+
+                if snapshot:
+                    if snapshot.volume_name != volume_name:
+                        context.set_details(
+                            messages.SnapshotWrongVolumeError_message.format(snapshot_name, snapshot.volume_name,
+                                                                             volume_name))
+                        context.set_code(grpc.StatusCode.ALREADY_EXISTS)
+                        return csi_pb2.CreateSnapshotResponse()
+                else:
+                    logger.debug(
+                        "Snapshot doesn't exist. Creating a new snapshot {0} from volume {1}".format(snapshot_name,
+                                                                                                     volume_name))
+                    snapshot = array_mediator.create_snapshot(snapshot_name, volume_name)
+
+                logger.debug("generating create snapshot response")
+                res = utils.generate_csi_create_snapshot_response(snapshot, source_volume_id)
+                logger.info("finished create snapshot")
+                return res
+        except (controller_errors.IllegalObjectName, controller_errors.VolumeNotFoundError) as ex:
+            context.set_details(ex.message)
+            context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
+            return csi_pb2.CreateSnapshotResponse()
+        except controller_errors.PermissionDeniedError as ex:
+            context.set_code(grpc.StatusCode.PERMISSION_DENIED)
+            context.set_details(ex)
+            return csi_pb2.CreateSnapshotResponse()
+        except (controller_errors.SnapshotAlreadyExists,
+                controller_errors.SnapshotNameBelongsToVolumeError) as ex:
```

Comment:
this is not a snapshot, so I'm 100% sure we can't use ALREADY_EXISTS for this.
[the spec is very clear about it](https://github.com/container-storage-interface/spec/blob/v1.2.0/spec.md#createsnapshot-errors)

Reply:
will return INTERNAL now